lib/cdo/video/youtube.rb

```diff
@@ -0,0 +1,66 @@
+# This class downloads and processes a YouTube video by ID,
+# making it available for viewing through Code.org's fallback video player.
+
+# Dependency requirements:
+# viddl-rb (`gem install viddl-rb`) - for downloading the video
+# ffmpeg (`brew install ffmpeg` / `apt-get install ffmpeg`) - for transcoding local videos
+# s3_access_key_id and s3_secret_access_key defined (e.g. in `locals.yml`)
+
+require 'cdo/aws/s3'
+require 'tmpdir'
+require 'open-uri'
+
+class Youtube
+
+  VIDEO_BUCKET = 'videos.code.org'
+
+  # Process a video with a youtube identifier.
+  # If filename is provided, transcode the local file with ffmpeg.
+  # If no filename is provided, download the transcoded video from YouTube.
+  #
+  # When downloading from YouTube, an HTTP head request will first check the absence of the file.
+  # If `force`==true, the head request will be skipped.
+  def self.process(id, filename=nil, force=false)
+    if filename.nil? && !force
+      require 'httparty'
+      thumbnail_url = "https://#{VIDEO_BUCKET}/youtube/#{id}.jpg"
+      response = HTTParty.head(thumbnail_url).response
+      if response.is_a? ::Net::HTTPSuccess
```

Comment:
There shouldn't be any harm if a spurious failure causes the filter to re-process the video. I made this condition intentionally narrow so that the filter would err on the side of caution in the case of uncertainty.